File: python/camera_ui_sdk/internal/sensor_rpc.py
from __future__ import annotations

from collections.abc import Callable
from typing import TYPE_CHECKING, Any, NotRequired, TypeAlias, TypedDict

if TYPE_CHECKING:
    from ..sensor.audio import AudioProperty
    from ..sensor.base import SensorCategory, SensorType
    from ..sensor.battery import BatteryCapability, BatteryProperty
    from ..sensor.classifier import ClassifierProperty
    from ..sensor.contact import ContactProperty
    from ..sensor.doorbell import DoorbellProperty
    from ..sensor.face import FaceProperty
    from ..sensor.garage import GarageProperty
    from ..sensor.humidity import HumidityProperty
    from ..sensor.leak import LeakProperty
    from ..sensor.license_plate import LicensePlateProperty
    from ..sensor.light import LightCapability, LightProperty
    from ..sensor.motion import MotionProperty
    from ..sensor.object import ObjectProperty
    from ..sensor.occupancy import OccupancyProperty
    from ..sensor.ptz import PTZCapability, PTZProperty
    from ..sensor.security_system import SecuritySystemProperty
    from ..sensor.siren import SirenCapability, SirenProperty
    from ..sensor.smoke import SmokeProperty
    from ..sensor.spec import ModelSpec
    from ..sensor.switch import SwitchProperty
    from ..sensor.temperature import TemperatureProperty

    SensorPropertyType: TypeAlias = (
        AudioProperty
        | BatteryProperty
        | ClassifierProperty
        | ContactProperty
        | DoorbellProperty
        | FaceProperty
        | GarageProperty
        | HumidityProperty
        | LeakProperty
        | LicensePlateProperty
        | LightProperty
        | MotionProperty
        | ObjectProperty
        | OccupancyProperty
        | PTZProperty
        | SecuritySystemProperty
        | SirenProperty
        | SmokeProperty
        | SwitchProperty
        | TemperatureProperty
    )

    SensorCapability: TypeAlias = PTZCapability | LightCapability | SirenCapability | BatteryCapability


class PropertyChangedEvent(TypedDict):
    """Emitted when a sensor property value changes."""

    sensorId: str
    """Sensor that changed."""
    sensorType: SensorType
    """Type of the sensor that changed."""
    property: str
    """Property key that changed."""
    value: object
    """New value."""
    previousValue: NotRequired[object]
    """Value before the change, absent on the first write."""
    timestamp: int
    """Change time in epoch milliseconds."""


PropertyUpdateFn = Callable[[dict[str, Any]], None]
"""Receives a partial-state delta, one call per ``_write_state``."""

CapabilityUpdateFn = Callable[[list[str]], None]
"""Receives the full capability list whenever it changes."""


class SensorJSON(TypedDict):
    """JSON-serializable representation of a sensor for RPC transport."""

    id: str
    """Sensor ID."""
    type: SensorType
    """Sensor type."""
    name: str
    """Internal name."""
    displayName: str
    """Name shown in the UI."""
    category: SensorCategory
    """Category the sensor belongs to."""
    nativeId: NotRequired[str]
    """Device ID assigned by the plugin."""
    pluginId: NotRequired[str]
    """Plugin that owns the sensor."""
    properties: dict[str, Any]
    """Current property values."""
    capabilities: NotRequired[list[str]]
    """Capability keys the sensor reports."""
    requiresFrames: NotRequired[bool]
    """Sensor needs a frame feed to work."""
    modelSpec: NotRequired[ModelSpec]
    """Model the sensor runs, for ML-backed sensors."""

File: node/src/internal/sensor-rpc.ts
import type { SensorCategory, SensorPropertyType, SensorType } from '../sensor/base.js';
import type { ModelSpec } from '../sensor/spec.js';

/** Emitted when a sensor property value changes. */
export interface PropertyChangedEvent {
  /** Sensor that changed. */
  sensorId: string;
  /** Type of the sensor that changed. */
  sensorType: SensorType;
  /** Property key that changed. */
  property: SensorPropertyType;
  /** New value. */
  value: unknown;
  /** Value before the change, absent on the first write. */
  previousValue?: unknown;
  /** Change time in epoch milliseconds. */
  timestamp: number;
}

/**
 * Receives a partial-state delta, one call per `_writeState`.
 *
 * @internal
 */
export type PropertyUpdateFn = (properties: Record<string, unknown>) => void;

/**
 * Receives the full capability list whenever it changes.
 *
 * @internal
 */
export type CapabilityUpdateFn = (capabilities: string[]) => void;

/** JSON-serializable representation of a sensor for RPC transport. */
export interface SensorJSON {
  /** Sensor ID. */
  id: string;
  /** Sensor type. */
  type: SensorType;
  /** Internal name. */
  name: string;
  /** Name shown in the UI. */
  displayName: string;
  /** Category the sensor belongs to. */
  category: SensorCategory;
  /** Device ID assigned by the plugin. */
  nativeId?: string;
  /** Plugin that owns the sensor. */
  pluginId?: string;
  /** Current property values. */
  properties: Record<string, unknown>;
  /** Capability keys the sensor reports. */
  capabilities?: string[];
  /** Sensor needs a frame feed to work. */
  requiresFrames?: boolean;
  /** Model the sensor runs, for ML-backed sensors. */
  modelSpec?: ModelSpec;
}
